### ISSUE TITLE
resixe remote screen to window

### ISSUE THREAD

**Issue by pcause** [TigerVNC/tigervnc#274]
I have this option set.  I'd have expected that this would scale the remote screen to the local window.  I get scroll bars instead.  Not sure if this is a bug, but if not than I'd make the request for an enhancement to scale the remote screen to the local window.

**Comment by CendioOssman** [comment]
The resize option does not scale, no. It tells the server to change the size of the session. Many servers do no support this (in fact, I think only TigerVNC does).

**Comment by pcause** [comment]
thanks.  yes, what I am looking for is having the remote screen, whatever its size, scaled to fit the client window.  Just like an image scales when you shrink/expand a window.  Thanks for changing title/.status

**Comment by dcommander** [comment]
I'm not trying to troll this thread, but I wanted to point out that the TurboVNC Server (2.0 and later) also supports the remote desktop resizing feature, which is a useful sanity check if any compatibility problems are suspected.

As far as scaling, our viewers support that, and they can be used with the TigerVNC Server.

**Comment by oneru** [comment]
@CendioOssman This seems to work perfectly when the server is running on Linux.  Is it possible to make this work when a Windows machine is running the server component?

**Comment by CendioOssman** [comment]
Possibly. But probably only to the fixed resolutions that are already available.

We don't have anyone working on WinVNC though, so it's not really getting much in the way of improvements. You could add an enhancement request though, and maybe put a bounty on it.

**Comment by thanatos** [comment]
This is so needed. When I connect to my home laptop w/ TigerVNC on my MBP, TigerVNC displays each remote pixel as 2x2 (2x zoom) on the MBP, so the screen is hugely zoomed in and the scrollbaring is just painful. (It seems like some sort of "retina" thing, as if it doesn't think my home laptop — which has a HiDPI screen — has a HiDPI screen.)

**Comment by fnJeff** [comment]
Agree with the "so needed" comment above.  Working from home will be next to impossible since my monitors are smaller than at work.  Changing the server resolution only leaves my carefully placed windows half off the screen.  Local client scaling would solve this problem.

If I possessed the software skills to do so I would, but I'm a lowly EE whose company IT only supports TigerVNC for all engineering work.  So far everything works perfectly, except this one thing.

Have any bounties been offered on this feature?  How much would it take to make it happen?

**Comment by CendioOssman** [comment]
No bounty yet, no. And unfortunately I have no idea what it would take to pique someone's interest. You'll just have to start and see if you get some attention. :)

Note that the bountysource integration is currently broken, so it won't get properly flagged here in github.

**Comment by fnJeff** [comment]
The reason I asked what it would take is because I had the crazy idea of seeing if there was any possibility of getting my company to pay for it.  Not sure how likely that would even be, but without a price tag on getting it done for sure, I can't really even ask the question.

This is probably a stupid question, but I think there are other VNC clients that already support this feature.  Do different open source versions of VNC ever share that kind of code with each other?  I think it was mentioned above that TurboVNC supports client scaling.  I tried using that with whatever server our IT is running (I'm not sure what it is, but I don't think it's TigerVNC server, and I'm not sure how I find out what it is) and it didn't work.  Something about unsupported or incorrect authentication or encryption modes.  I also tried TightVNC and RealVNC viewers, similar issues.  TigerVNC seems to be the only one that supports whatever form of authentication IT has setup on the server, and is the only one that IT will officially support anyway.

But without client scaling, it's practically unusable for my setup when changing between 2 locations with different resolution capabilities.  Any help, guidance, or ideas on how it would be possible to get this feature added would be highly appreciated.

**Comment by bphinz** [comment]
Like DRC's TurboVNC client, the TigerVNC Java client also supports client-side scaling.

**Comment by fnJeff** [comment]
So when I disable (uncheck) the option "Resize remote session to the local window" and then change the size of the client window, it does not scale.  It crops and gives me scroll bars.  If it supported client side scaling I would expect an option to select if that's what is desired, as well as the option to lock the aspect ratio when doing so.

I simply downloaded vncviewer.exe and run that on windows.  Maybe I'm not using the Java client you specifically referred to?

**Comment by bphinz** [comment]
That's the native client.  The Java client has the options you are expecting and can be downloaded here:

https://bintray.com/tigervnc/stable/download_file?file_path=VncViewer-1.8.0.jar

Then just double click it to run (requires you to have Java 6 or higher installed).

**Comment by bphinz** [comment]
You can also download the latest nightly build from the link on the tigervnc.org home page.  There's some fixes in the latest versions related to cursors on Windows.

**Comment by fnJeff** [comment]
Ok, this may sound weird, but I think I love you.  :P

I absolutely can't thank you enough for pointing me in the right direction.  The Java client works perfectly and does exactly what I need.

You may have saved my job at this company... thanks again!!!!!

**Comment by bphinz** [comment]
Happy to help, sorry that I didn't see this issue sooner!

**Comment by fnJeff** [comment]
No worries, just glad you saw it!  I'm already in the process of compiling directions to everyone in my group and I'm going to tell IT about it as well.  Hopefully the native client will have the same feature one day, but I'm happy for now.

Had to do a little searching on how to pin a jar to the taskbar and make it's icon the TigerVNC icon, but was a piece of cake to do that.

Can't say thanks enough times... :)

**Comment by bphinz** [comment]
You might want to look into Java WebStart, it's a great way to deploy jar files so that they look & feel like native apps.  More importantly, it allows updates to be pushed out remotely.

**Comment by fnJeff** [comment]
I'll certainly mention it to IT when I tell them about the Java client.  I'm just a lowly user that refused to deal with something that was annoying me.  :P

**Comment by fnJeff** [comment]
I knew I got a little to happy a little too soon and jinxed myself.

The Java client is not handling certain key combinations correctly.  I use xemacs extensively and can't seem to do combinations of Shift, Ctrl, and Alt.  For instance, Shift-Ctrl-Alt L is used to toggle between buffers.  After switch to the Java client, xemacs only seems to see Shift L.  This is actually a more serious issue than I started with... bummer.

I found a somewhat old issue about this that seemed to have to do with FLTK being out of date.  Don't think that applies to the Java client though...

Help?  New issue?

**Comment by bphinz** [comment]
I'll have to look and see if I can reproduce this, but yes this should be considered a separate issue.

**Comment by amonott** [comment]
We are quite interested in scaling with the native client, mainly to be able to reduce VNC traffic. We would be willing to fund this at least partially. The Java client is no option for us.

I have already looked into the code for a while, but the dependencies between the layers seem too complicated for a quick solution. A fixed scaling factor through a parameter, e.g. 2x2, would be great - then resizing could still work, just that the output is always scaled.

**Comment by vv7v7** [comment]
bump

**Comment by wzab** [comment]
I fully agree that a constant scaling factor selected with the command line parameter (e.g. --scale 3x3) would be perfect. I need to access the same VNC session both from the old display with 1280x1024 resolution end from the new 4k display. Without that feature it is really a nightmare.

**Comment by wzab** [comment]
I have just found "gvncviewer" available for my Debian system. It supports client-side scaling.

**Comment by lmajewski** [comment]
If I may ask - is there any update on this issue on TigerVNC 1.9.0?

I'm using debian's ```tigervnc-scraping-server``` package to get access to my remote's PC (DISPLAY=:0) via VPN (x0vncserver).

**Comment by CendioOssman** [comment]
Yes and no. The graphics handling has gotten some upgrades that should make this easier. But no one has looked specifically at scaling as far as I know.

**Comment by rocket-pig** [comment]
for ubuntu, 'vinagre' does window scaling a treat. (sudo apt install vinagre should do it)

makes iphone/other retina displays usable: 
![image](https://user-images.githubusercontent.com/24867505/71294485-816b1300-232d-11ea-94f2-a1693b8e9ef0.png)

**Comment by DocMAX** [comment]
Why is client side scaling in the Java version but not in the regular one??!?!?!?!?!? Are you kidding us?!

**Comment by bphinz** [comment]
On Sat, Apr 3, 2021 at 2:48 PM DocMAX ***@***.***> wrote:

> Why is client side scaling in the Java version but not in the regular
> one??!?!?!?!?!? Are you kidding us?
>
Because the current stable version of Fltk doesn’t support it.  Scaling is
supposed to be available in Fltk v1.4.0

> --
Sent from Gmail Mobile

**Comment by michaelcadilhac** [comment]
Also interested in that feature; when teaching, I just vnc to my work computer on a vncserver with low resolution, and scale it (on realvnc) so that students in the classroom can see an embiggened version of my screen.  Other options (using GDK_SCALE and similar) are never satisfactory, but this is.

**Comment by fottsia** [comment]
Same here, it would be great to have this feature as not all student's screens have the same resolution as the teacher's screen. Currently when scaling is needed, we resort to ssvnc or realvnc.

**Comment by DocMAX** [comment]
but with remina it's not gpu accelerated. tigervnc <-> tigervnc server is much faster.
1.13.1 and STILL no local resize option!!!

**Comment by CendioOssman** [comment]
If both server and client are TigerVNC, then the session should be able to resize dynamically to the client, rather than scaling. Is that not working for you?

**Comment by DocMAX** [comment]
Absolutely not right for me. I also have x11vnc running as i need DRM3 sometimes. (Or tell me if i can make DRM3 work in vncserver.) Target screen is 1080p and my monitor is 1080p. I dont want to run fullscreen so i need a window mode with scaled image.

**Comment by pcause** [comment]
I also don't want to run full screen, I use wayvnc, so agree client scaling please.

**Comment by DocMAX** [comment]
i wonder what is packet in the 34mb large exe file as it is not even capable of local scaling. (tightvnc only has 1mb and supports scaling!)

**Comment by romanrm** [comment]
In my case I connect to the same 1080p VNC session from a 1080p laptop, and from a 4K screen desktop. While it is perfect as-is on the laptop, connecting from the desktop it is more comfortable to view it at 1.5x scaling. Thanks

**Comment by alkisg** [comment]
> Scaling is supposed to be available in Fltk v1.4.0

I'm not familiar with FLTK at all. Which function do we care about, is it [scale](https://www.fltk.org/doc-1.3/classFl__Shared__Image.html#a1026830f0b7384269a5b8a304a438f52)?
That one seems to be available since FLTK 1.3.4, which corresponds to e.g. Ubuntu 18.04.
But even in Ubuntu 22.04 I don't get scaled contents when I resize the client VNC window (using e.g. x11vnc as the server).

Are we still waiting for more upstream FLTK changes before this can be implemented?
I'm not looking for server-side resizing or encoding. I only care about the image to fit the window when I resize the VNC client...

**Comment by CendioOssman** [comment]
We are not waiting for FLTK for this one, no. That's for high DPI support.

What's needed here is that someone looks at the current rendering pipeline and figures out how to efficiently add scaling to that. Basically, the "Surface" classes in vncviewer.

**Comment by cyfex** [comment]
I think this function is much easier than the current behaviour of window resizing---telling the VNC server to change geometry. Scaling can be done by just resizing the frames the client receives. That's what I think, at least.

**Comment by Blu3cr0ss** [comment]
issue is 7 years old and still not solved?

**Comment by DocMAX** [comment]
nope

**Comment by telychko** [comment]
Hey, there is a [draft pull request](https://github.com/TigerVNC/tigervnc/pull/1672) for x11 implementation

**Comment by bphinz** [comment]
I came across [libswscale](https://ffmpeg.org/libswscale.html) just a few days ago and from a cursory look it seems like it would be a minimally invasive way to scale the native pixel buffers. The nice thing there would be that it should be cross-platform. I have no idea what the performance is like but IMO scaling should be expected to come at a cost.

**Comment by telychko** [comment]
> I came across [libswscale](https://ffmpeg.org/libswscale.html) just a few days ago and from a cursory look it seems like it would be a minimally invasive way to scale the native pixel buffers. The nice thing there would be that it should be cross-platform. I have no idea what the performance is like but IMO scaling should be expected to come at a cost.

As I understood correctly, there is used low level rendering API for each platform. If we would use any cross-platform library for rendering here, we could have scaling and zooming a long time ago.

**Comment by bphinz** [comment]
I was looking at [this function](https://ffmpeg.org/doxygen/0.5/imgresample_8c.html#7bcbec92541f139023093e37a7de04de) which could be used to create a scaled copy of the pixel buffer.  I don't see why anything should need to change with regards to rendering because the changes you've already made in the PR would setup the surfaces to correspond to the size of the scaled buffer.  This is basically the same thing that I did to add scaling to the Java viewer.

**Comment by CendioOssman** [comment]
ffmpeg is problematic since distros often exclude it for patent reasons. libswscale is likely fine in that regard, but it might still be excluded since it is part of ffmpeg. We'll need to check that it is widely available before we rely on it.

I assume it is also a pure software implementation? We should try to leverage hardware acceleration first, so the scaling doesn't become a big bottleneck.

**Comment by DocMAX** [comment]
how many years still to come until local scaling is supported. why is it so hard to do?

**Comment by telychko** [comment]
What is the roadmap for implementing this feature?
Is it OK if it would be implemented for X11 only?

How to integrate this feature with server side resize?
1. It can be an option to choice (server side resize or client size resize)
2. Client size resize works only if server side resize don't
In my opinion option 1 is better as it gives choice to user and easier to implement.

**Comment by CendioOssman** [comment]
> What is the roadmap for implementing this feature? Is it OK if it would be implemented for X11 only?
> 

Not really. We want to maintain feature parity between our different platforms.

> How to integrate this feature with server side resize?
> 
> 1. It can be an option to choice (server side resize or client size resize)
> 
> 2. Client size resize works only if server side resize don't
>        In my opinion option 1 is better as it gives choice to user and easier to implement.

We should avoid bothering the user with settings if we can. So I would prefer we try with everything always enabled and see what feedback we get.

**Comment by AbandonedCart** [comment]
I was directed here by a lot of reviews hailing this as one of the best viewers with a huge community of support. 

Setup was easy enough after I figured out which of the practically indistinguishable binaries was the installed viewer. I connected in Window mode, but the resolution on the Mac is significantly larger than my Windows machine. There are some universally understood standards and one is that setting fullscreen would scale to the size of machine. 

Instead, I had to fight my way back out of the app to discover we are 7 years info this being a known flaw that has been explained by "not bothering the user with settings." Which do you think is more frustrating. An extra checkbox or fighting your way out of the app so you could uninstall it?

... I think you need to reevaluate your priorities on this one.

**Comment by samhed** [comment]
The reason that this still isn't available in TigerVNC is **not** that we want to avoid additional settings.

**Comment by adrian-moisa** [comment]
Why close the issue. We really need scaling. It's such an essential feature. I'm doing a lot of times remote E2E tests for the pipeline and I need tigerVNC to connect to wayVNC. It was the only one that worked with the required auth settings. I miss the scaling feature from RealVNC. It was super useful. I often made the window smaller to keep it on the side to watch for tests progress. Please do consider implementing scaling.

**Comment by blooumpy** [comment]
> Why close the issue. We really need scaling. It's such an essential feature. I'm doing a lot of times remote E2E tests for the pipeline and I need tigerVNC to connect to wayVNC. It was the only one that worked with the required auth settings. I miss the scaling feature from RealVNC. It was super useful. I often made the window smaller to keep it on the side to watch for tests progress. Please do consider implementing scaling.

I agree - the other vnc clients (tightvnc, realvnc) have it, odd that it's not available here.  Drag a window corner and desktop will resize to fit to the window size.  I don't have the skillset to assist with implementation nor do I wish to sound entitled - but I really do think it ought to be a standard feature.

**Comment by samhed** [comment]
> Why close the issue.

This issue is not closed. Please don't hesitate to submit a pull request.

**Comment by JavaScythe** [comment]
Would still like to see this

**Comment by mdevaev** [comment]
@CendioOssman what do you think about software scaling using FLTK 1.4? My users also complain that I recommend TigerVNC, but there is no scaling. This is really bad considering the proliferation of 2K and 4K screens, and it seems to me that software scaling should work quickly on modern hardware.

**Comment by CendioOssman** [comment]
The session is drawn using low-level calls, not using FLTK. So I'm afraid FLTK 1.4 will not help out here.

**Comment by mdevaev** [comment]
Doesn't resizing using `Fl_RGB_Image::scale()` **before** `Surface::update()` (and scaling the surface to the required size) solve the problem?

**Comment by jzvikart** [comment]
> Would still like to see this

Me too, it is an essential feature.

**Comment by namquang93** [comment]
Yeah it's almost impossible to remote desktop to a Mac that has a high resolution monitor from a Windows PC with lower resolution monitor.

**Comment by edwyn-amador** [comment]
Any update on this?

**Comment by JavaScythe** [comment]
Happy decade anniversary Cendio

**Comment by FunctionPoint** [comment]
Me too want this :)
TigerVNC is useless without it when connecting from FHD clients to Retina hosts.

**Comment by auouymous** [comment]
The patch by @ktmizugaki has been working great, other than the low quality which is fine for most cases. Hopefully I or someone else will get time soon to implement filtering, and then get it merged.

**Comment by heartacker** [comment]
how to make this possible?

**Comment by FajeSu** [comment]
@CendioOssman 
Scaling is a very necessary thing. Please implement it!🙏🏻🙏🏻🙏🏻

**Comment by haranloga** [comment]
I am frustrated by the lack of native, fluid client-side image scaling when resizing the TigerVNC Viewer window on a Windows client connected to a Linux server.

Currently, TigerVNC tries to solve resizing by modifying the remote server's actual resolution on the fly using X11 RANDR extensions.

HiDPI/OS Scaling Conflicts: When the Windows host uses OS-level scaling (e.g., 125% or 150%), the window sizing calculations fail. This results in massive cursor offsets, blurry text layouts, or unnecessary scrollbars rather than an optimized window view.

Rigid Server Layouts: If the Linux backend or desktop environment blocks dynamic resolution shifts, the user is forced into scrollbar navigation with no easy way to simply shrink or stretch the desktop bitmap to fit a smaller monitor region.

I would like an explicit, easily accessible "Autoscale Desktop to Window" toggle built directly into the client GUI (under the Options menu or as a hotkey).

When enabled:The native TigerVNC viewer should locally stretch or shrink the incoming remote desktop bitmap to cleanly fit the boundaries of the local window layout.It should perform this scaling on the client side without attempting to send RANDR structural alteration requests to the Linux server.

With the advent of modern AI coding assistants, refactoring the viewer UI and integrating basic image-scaling algorithms into the rendering loop should be significantly easier to implement. AI can streamline rewriting the scaling logic, bridging the UI gap, and generating the necessary boilerplate code to natively map the -ScalingPercentage=Auto backend behavior straight to a front-end toggle.

**Comment by pcause** [comment]
I am the original poster.  Switched to using remmina.

**Comment by haranloga** [comment]
does remmina support wayvnc , wayland

**Comment by pcause** [comment]
I use remmina to access a raspberry pi 500 running sway headless.  wayvnc is on the pi 500 reminna on fedora x86_64
